### PR TITLE
inline callbacks

### PR DESCRIPTION
**Pull request by sigatrev**
This commit adds tracking of up to 1 function passed as an argument to any given callsite, and provides the neccessary data to the JIT to inline it. The inliner is able to follow the single-def chain up multiple functions to find the approriate data, allowing inling to occur at megamorphic in cases where the function is a callback. This will allow more efficient javascript implementations of built-ins like filter and sort.

### CODE REVIEW

File: lib/JITIDL/JITTypes.h [hunk 1]
```diff
@@ -641,11 +652,14 @@ typedef struct FunctionJITTimeDataIDL
 
     unsigned int inlineeCount;
     unsigned int ldFldInlineeCount;
+    unsigned int callbackInlineeCount;
```
Comment (sethbrenith) [inline lib/JITIDL/JITTypes.h]: Could we put this int down in the X64_PAD4 space after callbackInlinees?

File: lib/Backend/InliningDecider.cpp [hunk 2]
```diff
@@ -121,7 +132,17 @@ Js::FunctionInfo *InliningDecider::InlineCallSite(Js::FunctionBody *const inline
...
+    }
+    return nullptr;
+}
+
+Js::FunctionInfo * InliningDecider::InlineCallback(Js::FunctionBody *const inliner, const Js::ProfileId profiledCallSiteId, uint recursiveInlineDepth)
+{
+    Js::FunctionInfo * functionInfo = GetCallSiteCallbackInfo(inliner, profiledCallSiteId);
+    if (functionInfo)
+    {
+        return Inline(inliner, functionInfo, false, false, true, GetConstantArgInfo(inliner, profiledCallSiteId), profiledCallSiteId, recursiveInlineDepth, true);
```
Comment (sethbrenith) [inline lib/Backend/InliningDecider.cpp]: >false [](start = 45, length = 5)

How do we guarantee at this point that it's not a constructor call?

File: lib/Backend/InliningDecider.h [hunk 3]
```diff
@@ -105,6 +107,10 @@ class InliningDecider
     Output::Print(__VA_ARGS__); \
     Output::Flush(); \
     }
+#define INLINE_TRACE_AND_TESTTRACE(...) \
```
Comment (sethbrenith) [inline lib/Backend/InliningDecider.h]: Thanks for cleaning this up :)

File: test/inlining/InlineCallbacks.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
```
Comment (sethbrenith) [inline test/inlining/InlineCallbacks.js]: Are there other configurations that might be worth testing? Maybe passing the callback through multiple layers like `function Dispatch2(f) { Dispatch(f); }` or a callback that calls things like `function Dispatch3(f) { f(Foo); }` where we call `Dispatch3(Dispatch)` after the warmup?

Comment (sethbrenith) [reply test/inlining/InlineCallbacks.js]: or maybe a case where the callback isn't the first param

---
In reply to: [185833307](https://github.com/Microsoft/ChakraCore/pull/5081#discussion_r185833307) [](ancestors = 185833307)

File: lib/Backend/FunctionCodeGenJitTimeData.h [hunk 5]
```diff
@@ -46,6 +47,9 @@ namespace Js
         // accurate count.
         Field(uint) ldFldInlineeCount;
 
+        // Number of functions passed as arguments to be inlined.
+        Field(uint) callbackInlineeCount;
```
Comment (sethbrenith) [inline lib/Backend/FunctionCodeGenJitTimeData.h]: We currently have four unused bytes after globalObjTypeSpecFldInfoCount; could this go up there?

Comment (sigatrev) [reply lib/Backend/FunctionCodeGenJitTimeData.h]: This member isn't actually needed. I'm going to remove it.

File: lib/Backend/InliningDecider.cpp (hunk 2 above)
Comment (sigatrev) [reply lib/Backend/InliningDecider.cpp]: constructor call is specific to the callsite, not the function. It is set true for calls of the form `new foo()`. At the point callback info is collected, we don't know how the callsite will use the callback function. The result of guessing wrong here is either failing to serialize data that could have been used, or serializing data that won't be used, but neither case is fatal

File: lib/Runtime/Language/DynamicProfileInfo.h [hunk 6]
```diff
@@ -84,6 +84,31 @@ namespace Js
...
+    {
+        bool CanInlineCallback(Js::ArgSlot argIndex)
+        {
+            return canInlineCallback && argNumber == argIndex;
+        }
+
+        // False if there is more than one ArgIn that is a function object or a function object with arg number greater than MaxInlineeArgoutCount
+        Field(uint8) canInlineCallback : 1;
+
+        Field(uint8) isConstructor : 1;
```
Comment (sethbrenith) [inline lib/Runtime/Language/DynamicProfileInfo.h]: unused?

Comment (sigatrev) [reply lib/Runtime/Language/DynamicProfileInfo.h]: I added that before looking into what/how that tracks. thanks for catching that.

**Pull request merged**